File: book/lib/services/notebook_service.dart
import '../models/notebook.dart';
import '../models/cell.dart';
import 'api_client.dart';

class NotebookService {
  final ApiClient _api;

  NotebookService({ApiClient? api}) : _api = api ?? apiClient;

  Future<List<Notebook>> list() async {
    try {
      final response = await _api.getList('/api/notebooks');
      return response.map((n) => _notebookFromJson(n as Map<String, dynamic>)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<Notebook?> get(String id) async {
    try {
      final response = await _api.get('/api/notebooks/$id');
      return _notebookFromJson(response);
    } catch (e) {
      return null;
    }
  }

  Future<Notebook?> create(String name, {String? kernelName}) async {
    try {
      final response = await _api.post('/api/notebooks', {
        'name': name,
        if (kernelName != null) 'kernel_name': kernelName,
      });
      return _notebookFromJson(response);
    } catch (e) {
      return null;
    }
  }

  Future<Notebook?> update(String id, {String? name, List<Cell>? cells}) async {
    try {
      final data = <String, dynamic>{};
      if (name != null) data['name'] = name;
      if (cells != null) {
        data['cells'] = cells.map((c) => {
          'id': c.id,
          'cell_type': c.cellType.name,
          'source': c.source,
          'outputs': c.outputs.map((o) => {
            'output_type': o.outputType,
            if (o.text != null) 'text': o.text,
            if (o.data != null) 'data': o.data,
            if (o.ename != null) 'ename': o.ename,
            if (o.evalue != null) 'evalue': o.evalue,
            if (o.traceback != null) 'traceback': o.traceback,
          }).toList(),
          if (c.executionCount != null) 'execution_count': c.executionCount,
          'status': c.status.name,
        }).toList();
      }
      final response = await _api.put('/api/notebooks/$id', data);
      return _notebookFromJson(response);
    } catch (e) {
      print('Error updating notebook: $e');
      return null;
    }
  }

  Future<bool> delete(String id) async {
    try {
      await _api.delete('/api/notebooks/$id');
      return true;
    } catch (e) {
      return false;
    }
  }

  Future<Cell?> addCell(String notebookId, CellType type, String source, {int? position}) async {
    try {
      final response = await _api.post('/api/notebooks/$notebookId/cells', {
        'cell_type': type.name,
        'source': source,
        if (position != null) 'position': position,
      });
      return _cellFromJson(response);
    } catch (e) {
      return null;
    }
  }

  Future<Cell?> updateCell(String notebookId, String cellId, {String? source}) async {
    try {
      final response = await _api.put('/api/notebooks/$notebookId/cells/$cellId', {
        if (source != null) 'source': source,
      });
      return _cellFromJson(response);
    } catch (e) {
      return null;
    }
  }

  Future<bool> deleteCell(String notebookId, String cellId) async {
    try {
      await _api.delete('/api/notebooks/$notebookId/cells/$cellId');
      return true;
    } catch (e) {
      return false;
    }
  }

  Future<String?> exportToPython(String notebookId) async {
    try {
      final response = await _api.get('/api/notebooks/$notebookId/export/python');
      return response as String?;
    } catch (e) {
      return null;
    }
  }

  Future<String?> exportToIpynb(String notebookId) async {
    try {
      final response = await _api.get('/api/notebooks/$notebookId/export/ipynb');
      if (response is Map) {
        return response.toString();
      }
      return response as String?;
    } catch (e) {
      return null;
    }
  }

  String getExportPythonUrl(String notebookId) {
    return '${ApiClient.baseUrl}/api/notebooks/$notebookId/export/python';
  }

  String getExportIpynbUrl(String notebookId) {
    return '${ApiClient.baseUrl}/api/notebooks/$notebookId/export/ipynb';
  }

  String getExportHtmlUrl(String notebookId) {
    return '${ApiClient.baseUrl}/api/notebooks/$notebookId/export/html';
  }

  Notebook _notebookFromJson(Map<String, dynamic> json) {
    return Notebook(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? 'Untitled',
      cells: (json['cells'] as List<dynamic>?)
              ?.map((c) => _cellFromJson(c as Map<String, dynamic>))
              .toList() ??
          [],
      kernelId: json['kernel_id'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : DateTime.now(),
    );
  }

  Cell _cellFromJson(Map<String, dynamic> json) {
    return Cell(
      id: json['id'] as String? ?? '',
      cellType: CellType.values.firstWhere(
        (t) => t.name == (json['cell_type'] ?? json['type']),
        orElse: () => CellType.code,
      ),
      source: json['source'] as String? ?? '',
      outputs: (json['outputs'] as List<dynamic>?)
              ?.map((o) => CellOutput.fromJson(o as Map<String, dynamic>))
              .toList() ??
          [],
      executionCount: json['execution_count'] as int?,
      status: CellStatus.values.firstWhere(
        (s) => s.name == json['status'],
        orElse: () => CellStatus.idle,
      ),
    );
  }
}

final notebookService = NotebookService();

File: notebook/src/services/notebookService.ts
// Notebook Service - CRUD operations for notebooks

import apiClient from "./apiClient";

export interface CellOutput {
  outputType: "stream" | "execute_result" | "display_data" | "error";
  text?: string;
  data?: Record<string, unknown>;
  ename?: string;
  evalue?: string;
  traceback?: string[];
}

export interface Cell {
  id: string;
  cellType: "code" | "markdown";
  source: string;
  outputs: CellOutput[];
  executionCount?: number;
  metadata?: Record<string, unknown>;
}

export interface Notebook {
  id: string;
  name: string;
  path: string;
  cells: Cell[];
  metadata: Record<string, unknown>;
  kernelSpec?: {
    name: string;
    displayName: string;
    language: string;
  };
  createdAt: Date;
  updatedAt: Date;
}

export interface CreateNotebookParams {
  name: string;
  path?: string;
  templateId?: string;
}

export interface NotebookTemplate {
  id: string;
  name: string;
  description: string;
  cells: Cell[];
  category: string;
}

interface NotebookResponse {
  id: string;
  name: string;
  path?: string;
  cells: Array<{
    id: string;
    cell_type: string;
    source: string;
    outputs: Array<{
      output_type: string;
      text?: string;
      data?: Record<string, unknown>;
      ename?: string;
      evalue?: string;
      traceback?: string[];
    }>;
    execution_count?: number;
    metadata?: Record<string, unknown>;
  }>;
  metadata: Record<string, unknown> & {
    created_at?: string;
    modified_at?: string;
  };
  kernel_spec?: {
    name: string;
    display_name: string;
    language: string;
  };
  kernel_id?: string | null;
  created_at?: string;
  updated_at?: string;
}

class NotebookService {
  private parseNotebook(data: NotebookResponse): Notebook {
    // Get dates from metadata or root level (backend may use either)
    const createdAt = data.created_at || data.metadata?.created_at || new Date().toISOString();
    const updatedAt = data.updated_at || data.metadata?.modified_at || createdAt;

    return {
      id: data.id,
      name: data.name,
      path: data.path || "",
      cells: (data.cells || []).map((cell) => ({
        id: cell.id,
        cellType: cell.cell_type as "code" | "markdown",
        source: cell.source,
        outputs: (cell.outputs || []).map((o) => ({
          outputType: o.output_type as CellOutput["outputType"],
          text: o.text,
          data: o.data,
          ename: o.ename,
          evalue: o.evalue,
          traceback: o.traceback,
        })),
        executionCount: cell.execution_count,
        metadata: cell.metadata,
      })),
      metadata: data.metadata,
      kernelSpec: data.kernel_spec
        ? {
            name: data.kernel_spec.name,
            displayName: data.kernel_spec.display_name,
            language: data.kernel_spec.language,
          }
        : undefined,
      createdAt: new Date(createdAt),
      updatedAt: new Date(updatedAt),
    };
  }

  private serializeNotebook(notebook: Partial<Notebook>): Record<string, unknown> {
    return {
      name: notebook.name,
      path: notebook.path,
      cells: notebook.cells?.map((cell) => ({
        id: cell.id,
        cell_type: cell.cellType,
        source: cell.source,
        outputs: cell.outputs.map((o) => ({
          output_type: o.outputType,
          text: o.text,
          data: o.data,
          ename: o.ename,
          evalue: o.evalue,
          traceback: o.traceback,
        })),
        execution_count: cell.executionCount,
        metadata: cell.metadata,
      })),
      metadata: notebook.metadata,
    };
  }

  async list(): Promise<Notebook[]> {
    const response = await apiClient.get<NotebookResponse[]>("/api/notebooks");
    return response.map((n) => this.parseNotebook(n));
  }

  async get(id: string): Promise<Notebook> {
    const response = await apiClient.get<NotebookResponse>(`/api/notebooks/${id}`);
    return this.parseNotebook(response);
  }

  async create(params: CreateNotebookParams): Promise<Notebook> {
    const response = await apiClient.post<NotebookResponse>("/api/notebooks", {
      name: params.name,
      path: params.path,
      template_id: params.templateId,
    });
    return this.parseNotebook(response);
  }

  async update(id: string, notebook: Partial<Notebook>): Promise<Notebook> {
    const response = await apiClient.put<NotebookResponse>(
      `/api/notebooks/${id}`,
      this.serializeNotebook(notebook)
    );
    return this.parseNotebook(response);
  }

  async delete(id: string): Promise<void> {
    await apiClient.delete(`/api/notebooks/${id}`);
  }

  async duplicate(id: string, newName?: string): Promise<Notebook> {
    const response = await apiClient.post<NotebookResponse>(`/api/notebooks/${id}/duplicate`, {
      name: newName,
    });
    return this.parseNotebook(response);
  }

  async rename(id: string, newName: string): Promise<Notebook> {
    const response = await apiClient.patch<NotebookResponse>(`/api/notebooks/${id}`, {
      name: newName,
    });
    return this.parseNotebook(response);
  }

  // Cell operations
  async addCell(notebookId: string, cellType: "code" | "markdown", afterCellId?: string): Promise<Cell> {
    const response = await apiClient.post<{
      id: string;
      cell_type: string;
      source: string;
      outputs: Array<unknown>;
      execution_count?: number;
      metadata?: Record<string, unknown>;
    }>(`/api/notebooks/${notebookId}/cells`, {
      cell_type: cellType,
      after_cell_id: afterCellId,
    });

    return {
      id: response.id,
      cellType: response.cell_type as "code" | "markdown",
      source: response.source,
      outputs: [],
      executionCount: response.execution_count,
      metadata: response.metadata,
    };
  }

  async updateCell(notebookId: string, cellId: string, source: string): Promise<void> {
    await apiClient.patch(`/api/notebooks/${notebookId}/cells/${cellId}`, {
      source,
    });
  }

  async deleteCell(notebookId: string, cellId: string): Promise<void> {
    await apiClient.delete(`/api/notebooks/${notebookId}/cells/${cellId}`);
  }

  async moveCells(notebookId: string, cellIds: string[], afterCellId?: string): Promise<void> {
    await apiClient.post(`/api/notebooks/${notebookId}/cells/move`, {
      cell_ids: cellIds,
      after_cell_id: afterCellId,
    });
  }

  // Import/Export
  async importFromFile(file: File): Promise<Notebook> {
    const response = await apiClient.uploadFile<NotebookResponse>("/api/notebooks/import", file);
    return this.parseNotebook(response);
  }

  async exportToIpynb(id: string): Promise<Blob> {
    const response = await fetch(`${apiClient.getBaseUrl()}/api/notebooks/${id}/export/ipynb`);
    return response.blob();
  }

  async exportToPython(id: string): Promise<string> {
    return apiClient.get<string>(`/api/notebooks/${id}/export/python`);
  }

  async exportToHtml(id: string): Promise<string> {
    return apiClient.get<string>(`/api/notebooks/${id}/export/html`);
  }

  // Templates
  async getTemplates(): Promise<NotebookTemplate[]> {
    return apiClient.get<NotebookTemplate[]>("/api/notebooks/templates");
  }

  async createFromTemplate(templateId: string, name: string): Promise<Notebook> {
    const response = await apiClient.post<NotebookResponse>("/api/notebooks", {
      name,
      template_id: templateId,
    });
    return this.parseNotebook(response);
  }

  // Clear all outputs
  async clearOutputs(id: string): Promise<void> {
    await apiClient.post(`/api/notebooks/${id}/clear-outputs`, {});
  }

  // Run all cells
  async runAll(id: string): Promise<void> {
    await apiClient.post(`/api/notebooks/${id}/run-all`, {});
  }
}

export const notebookService = new NotebookService();
export default notebookService;
